drivers/sensor/tsl2540/tsl2540_trigger.c

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright (c) 2022 T-Mobile USA, Inc.
+ * Copyright (c) 2023 T-Mobile USA, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "tsl2540.h"
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_DECLARE(tsl2540, CONFIG_SENSOR_LOG_LEVEL);
+
+static void tsl2540_handle_cb(struct tsl2540_data *data)
+{
+#if defined(CONFIG_TSL2540_TRIGGER_OWN_THREAD)
+	k_sem_give(&data->trig_sem);
+#elif defined(CONFIG_TSL2540_TRIGGER_GLOBAL_THREAD)
+	k_work_submit(&data->work);
+#endif
+}
+
+static void tsl2540_gpio_callback(const struct device *dev, struct gpio_callback *cb,
+				  uint32_t pin_mask)
+{
+	struct tsl2540_data *data = CONTAINER_OF(cb, struct tsl2540_data, gpio_cb);
+	const struct tsl2540_config *config = data->dev->config;
+
+	if ((pin_mask & BIT(config->int_gpio.pin)) == 0U) {
+		return;
+	}
+
+	tsl2540_handle_cb(data);
+}
+
+static void tsl2540_handle_int(const struct device *dev)
+{
+	int ret;
+	uint8_t status;
+
+	ret = tsl2540_reg_read(dev, TSL2540_REG_STATUS, &status);
+	if (ret) {
+		LOG_ERR("Could not read status register (%#x), errno: %d", TSL2540_REG_STATUS, ret);
+	} else {
+		if ((1 << 7) & status) { /* ASAT */
+			/*
+			 * TODO:
+			 * Implement a mechanism triggered by the over-saturation of one or
+			 * both input amplifiers to automatically adjust the input gain to
+			 * keep the amplifiers from saturating.
+			 */
+
+			LOG_ERR("Interrupt status(%#x): %#x: ASAT", TSL2540_REG_STATUS, status);
+		}
+		if ((1 << 3) & status) { /* CINT */
+			LOG_INF("Interrupt status(%#x): %#x: CINT", TSL2540_REG_STATUS, status);
+		}
+		if ((1 << 4) & status) { /* AINT */
+			struct tsl2540_data *data = dev->data;
+
+			LOG_INF("Interrupt status(%#x): %#x: AINT", TSL2540_REG_STATUS, status);
+			if (data->als_handler) {
+				/*
+				 * TODO: Consider fetching the illuminance data and making it
+				 * available to the application; for example:
+				 *
+				 *	k_sem_take(&data->sem, K_FOREVER);
+				 *	...
+				 *	k_sem_give(&data->sem);
+				 */
+
+				data->als_handler(
+					dev, &(struct sensor_trigger){.type = SENSOR_TRIG_THRESHOLD,
+								      .chan = SENSOR_CHAN_LIGHT});
```

Comment:
Need to pass back a pointer to the original trigger provided by the user instead of allocating a new trigger on the stack. See #55130